### DISCUSSION TITLE
Can tud_task() be called inside the USB ISR?

### DISCUSSION THREAD

**Discussion by silseva** [hathach/tinyusb#1376] [Q&A] [answered]
Can the periodic update function `tud_task()` be called inside the USB interrupt routine just after the `tud_interrupt_handler()` function?

Rationale for this is to avoid the need of having to call `tud_task()` periodically in the code main loop, which sometimes can be difficult to manage if blocking functions are present. By looking at the code it seems that doing so is safe when tinyusb is used without an RTOS and/or without mutexes. Also, other libraries like libopencm3 explicitly state that their update function can be called inside the USB ISR, meaning that is something viable in some cases.

Thanks,
Silvano.

**Comment by hathach** [comment, marked as answer]
there shouldn't be any issue if you don't use RTOS, and having usb irq as low priority. micropython call tud_task in isr as well.

**Reply by silseva** [reply]
Great, thank you!

Silvano.

**Reply by nachiketathakur** [reply]
I am using TinyUSB on RP2040. I have the same question, but I cannot find the `tud_int_handler()` in which I can go and place the `tud_task()` . So where can I find the `tud_int_handler()` in case of RP2040? @hathach

**Reply by hathach** [reply]
check out pico-sdk for how it install the usb irq handler.

**Comment by alberto-grl** [comment]
Directly calling tud_task() in the USB callback does not work for me. It gets stuck in `dcd_edpt_disable` at line
`while((usb_otg->GINTSTS & USB_OTG_GINTSTS_BOUTNAKEFF_Msk) == 0);`
I am not using any OS, on an STM32H723, CDC + UAC.

I made it work by making the call in a lower priority EXTI ISR, to let the magic in tud_task happen after the end of the USB interrupt.
Probably something in tud_task fires an USB interrupt that will not be serviced, being it already active. 
It would be nice to avoid the software interrupt, from this thread it seems that it was possible, once.

Having the periodic call triggered by the USB interrupt makes TinyUSB more similar to ST's USB stack, easier to integrate in existing projects.
It should be more efficient too. My test were driven by the finding that the audio loopback in the MSC+UAC example tud_task needs to be called at about 20 Khz. 

This is the code I'm using:

```
void OTG_HS_IRQHandler(void)
{
  /* USER CODE BEGIN OTG_HS_IRQn 0 */
			tud_int_handler(BOARD_DEVICE_RHPORT_NUM);
	               __HAL_GPIO_EXTI_GENERATE_SWIT(GPIO_PIN_13);  // Triggers a lower priority ISR
	              return;
  /* USER CODE END OTG_HS_IRQn 0 */
  HAL_PCD_IRQHandler(&hpcd_USB_OTG_HS);
  /* USER CODE BEGIN OTG_HS_IRQn 1 */
  /* USER CODE END OTG_HS_IRQn 1 */
}
```


```
void  HAL_GPIO_EXTI_Callback(uint16_t GPIO_Pin){
	if(GPIO_Pin == GPIO_PIN_13) {
		tud_task();
		audio_task();
		cdc_task();
	}
}
```


Regards

Alberto

**Reply by HiFiPhile** [reply]
> It gets stuck in dcd_edpt_disable at line
while((usb_otg->GINTSTS & USB_OTG_GINTSTS_BOUTNAKEFF_Msk) == 0);

It should be an issue of STM32H723, I always call `tud_task()` inside ISR on STM32L0/L1/F4, LPC43xx, LPC55, etc.

> making the call in a lower priority EXTI ISR

PendSV is made for "lower priority" or "deferred procedure call" usage, no need to use a hardware related interrupt.

```c
void SysTick_Handler(void)
{
    /* Time base tick */
    msTicks++;
    /* Trigger PendSV */
    SCB->ICSR = SCB_ICSR_PENDSVSET_Msk;
}

void USB1_IRQHandler(void)
{
    tud_int_handler(1);
    /* Trigger PendSV */
    SCB->ICSR = SCB_ICSR_PENDSVSET_Msk;
}

void PendSV_Handler(void)
{
    tud_task();
    /* Other deferred call, you can stay longer without mess with tick */
    DPC_Exec();
}
```

**Reply by hathach** [reply]
to be honest, I rarely call tud_task() within USB ISR, there may be a couple of gotcha. Though, if you do, make sure you use `tud_task_ext(timeout, true)`. I haven't actually tested this yet. But the in_isr indicator can be useful, e.g prevent blocking wait. Also depending port driver, some may assume it is always called in thread context, and could wait for some event e.g delay using SOF etc ... Overally, it is not well tested.

Thanks @HiFiPhile that is a very helpful tips

**Comment by alberto-grl** [comment]
Thanks to both.

> It should be an issue of STM32H723, I always call tud_task() inside ISR on STM32L0/L1/F4, LPC43xx, LPC55, etc.

Same thing happens on STM32H750. They are similar, next week I'll try with an STM32G4. 
But maybe there's something in the way peripherals are initialized, I have STMCube generated code. Apart from this tud_task call requirement everything works fine but who knows. 

> PendSV is made for "lower priority" or "deferred procedure call" usage, no need to use a hardware related interrupt.

Very useful, thanks. I would need more than one PendSV handler, EXTI has six INT lines, anyway this saves one pin. 

>  make sure you use tud_task_ext(timeout, true)

In device/usbd.c:

```
void tud_task_ext(uint32_t timeout_ms, bool in_isr)
{
  (void) in_isr; // not implemented yet
```

**Comment by Commotari** [comment]
i would not do that.
my understanding is that we have two parts of the USB stack:
* the fast ISR routines that need to empty the hardware buffer as fast as possible and return as fast as possible
* the slower part that can take the data and events that have been put into the queue from the fast part. it runs in the background.

putting the slow part (tud_task) into the end of the USB ISR has a disadvantage: if multiple usb packets are coming in, then maybe one of them is getting lost, because the USB ISR function is blocked by the slow tud_task.
it kind of sabotages the concept of the tinyusb stack to have a fast and slow part of the stack.

in my setup i have a millisecond interrupt that makes the timing, and from the normal application i call the tud-task approx. every 2ms.
( in my bootloader from my own mini-os, in my application from FreeRTOS )
and the USB ISR has a higher priority than other other interrupts (like adc, usart,...) to make sure that no incoming packet is getting lost.

hope that helps,
  WhyLee (Stefan Weilhartner)

**Comment by HiFiPhile** [comment]
> my understanding is that we have two parts of the USB stack:
the fast ISR routines that need to empty the hardware buffer as fast as possible and return as fast as possible
the slower part that can take the data and events that have been put into the queue from the fast part. it runs in the background.

That's correct, tinyusb use this approach to reduce interrupt blocking time.

> putting the slow part (tud_task) into the end of the USB ISR has a disadvantage: if multiple usb packets are coming in, then maybe one of them is getting lost, because the USB ISR function is blocked by the slow tud_task.
it kind of sabotages the concept of the tinyusb stack to have a fast and slow part of the stack.

I believe you have some fundamental misunderstanding of how USB transaction works. Only isochronous packet will lost if the receiver is not ready, for other type of transfer an acknowledge mechanism is in place.

In the case of isochronous transfer your timer approach won't help either.

You can read more about USB for example here:
https://www.beyondlogic.org/usbnutshell/usb4.shtml#Control
